langchain/tools/zapier.py

```diff
@@ -0,0 +1,197 @@
+# which you route your users through first
+
+actions = ZapierNLAWrapper().list()
+
+## step 1. gmail find email
+
+GMAIL_SEARCH_INSTRUCTIONS = "Grab the latest email from Bryan Helmig"
+
+def nla_gmail(inputs):
+    action = next((
+        a for a in actions if a["description"].startswith("Gmail: Find Email")
+    ), None)
+    data = ZapierNLARunAction(action_id=action["id"]).run(inputs["instructions"])
+    return {
+        "email_data": data
+    }
+gmail_chain = TransformChain(
+    input_variables=["instructions"],
+    output_variables=["email_data"],
+    transform=nla_gmail
+)
+
+## step 2. generate draft reply
+
+template = \"""You are an assisstant who drafts replies to an incoming email.
+Output draft reply in plain text (not JSON).
+
+Incoming email:
+{email_data}
+
+Draft email reply:\"""
+
+prompt_template = PromptTemplate(input_variables=["email_data"], template=template)
+reply_chain = LLMChain(llm=OpenAI(temperature=.7), prompt=prompt_template)
+
+## step 3. send draft reply via a slack direct message
+
+SLACK_HANDLE = "@knoop"
+
+def nla_slack(inputs):
+    action = next(
+        (a for a in actions if a["description"].startswith("Slack: Send Direct Message")
+    ), None)
+    instructions = f'Send this to {SLACK_HANDLE} in Slack: {inputs["draft_reply"]}'
+    return {"slack_data": ZapierNLARunAction(action_id=action["id"]).run(instructions)}
+slack_chain = TransformChain(
+    input_variables=["draft_reply"],
+    output_variables=["slack_data"],
+    transform=nla_slack
+)
+
+## finally, execute
+
+overall_chain = SimpleSequentialChain(
+    chains=[gmail_chain, reply_chain, slack_chain],
+    verbose=True
+)
+overall_chain.run(GMAIL_SEARCH_INSTRUCTIONS)
+
+
+```
+
+"""
+from typing import Optional
+
+from langchain.tools.base import BaseTool
+from langchain.utilities.zapier import ZapierNLAWrapper
+
+zapier_nla_base_desc = (
+    "A wrapper around Zapier NLA. "
+    "Can be used to call or retrieve data from 5k+ apps, 20k+ actions"
+    "on the Zapier platform."
+)
+
+
+class ZapierNLARunAction(BaseTool):
+    """
+    Args:
+        action_id: a specific action ID (from list actions) of the action to execute
+            (the set api_key must be associated with the action owner)
+        instructions: a natural language instruction string for using the action
+            (eg. "get the latest email from Mike Knoop" for "Gmail: find email" action)
+        params: a dict, optional. Any params provided will *override* AI guesses
+            from `instructions` (see "understanding the AI guessing flow" here:
+            https://nla.zapier.com/api/v1/dynamic/docs)
+    """
+
+    name = "Zapier NLA: Run Action"
+    description = zapier_nla_base_desc + (
+        "This tool will run a specified action and return a stringified JSON result "
+        " of the API call. The return result is guarenteed to be less than ~500 words "
+        " (350 tokens), safe to insert back into another LLM prompt."
```

Comment:
The description will be what the Agent uses to choose the tool. Additionally, the name must be unique for each tool. If we want several ZapierNLA tools, best to create them dynamically based on the available actions returned in `list`. This is what I did in the typescript integration https://github.com/hwchase17/langchainjs/pull/313 -- more than happy to make the changes here.